### PR TITLE
[CIRC-296, CIRC-297] Expanding notice tokens with user, loan, item and request info

### PR DESCRIPTION
**Pull request by dmytrotur**
## Purpose
Resolves: [CIRC-296](https://issues.folio.org/browse/CIRC-296), [CIRC-297](https://issues.folio.org/browse/CIRC-297)
Expand list of tokens available for for notice templates.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [x] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [x] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -629,7 +629,8 @@
             "proxiesfor.collection.get",
             "circulation.rules.notice-policy.get",
             "circulation-storage.patron-notice-policies.item.get",
-            "patron-notice.post"
+            "patron-notice.post",
+            "circulation-storage.cancellation-reasons.item.get"
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Please add the cancellation reasons interfaces to the set of dependencies for the module

File: src/main/java/org/folio/circulation/domain/Item.java [hunk 2]
```diff
@@ -5,22 +5,27 @@
 import static org.folio.circulation.domain.ItemStatus.CHECKED_OUT;
 import static org.folio.circulation.domain.ItemStatus.IN_TRANSIT;
 import static org.folio.circulation.domain.ItemStatus.MISSING;
+import static org.folio.circulation.domain.representations.ItemProperties.CONTRIBUTORS;
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/Item.java]: Contributors are part of an instance not an item, why is this using the item properties definition?

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (dmytrotur) [reply descriptors/ModuleDescriptor-template.json]: Done

File: src/main/java/org/folio/circulation/domain/Item.java (hunk 2 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/Item.java]: I missed that. Is it ok to create `InstanceProperties` for this key?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/Item.java]: Yes, that is ok, please do

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/Item.java]: Done

File: src/main/java/org/folio/circulation/domain/notice/NoticeContextUtil.java [hunk 3]
```diff
@@ -1,54 +1,157 @@
...
   }
 
-  public static JsonObject createLoanNoticeContext(Loan loan, DateTimeZone timeZone) {
-    return createNoticeContextFromItemAndPatron(loan.getItem(), loan.getUser())
-      .put("dueDate", loan.getDueDate().withZone(timeZone).toString());
+  public static JsonObject createLoanNoticeContext(
+    Loan loan, LoanPolicy loanPolicy, DateTimeZone timeZone) {
+
+    return new JsonObject()
+      .put(USER, createPatronContext(loan.getUser()))
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/NoticeContextUtil.java]: Given the grouping has been renamed from patron to user, would it be more consistent to rename the `createPatronContext` method to `createUserContext`?

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/notice/NoticeContextUtil.java]: Done

**Comment by dmytrotur** [comment]
I tested tokens functionality under with limited permissions and found several missing permissions. Added them in the last commit

File: src/test/java/api/support/fixtures/NoticeTokens.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+public class NoticeTokens {
+
+  private NoticeTokens() {
+  }
+
+  public static final List<String> EXPECTED_USER_TOKENS = Arrays.asList(
+    "user.firstName", "user.lastName", "user.barcode");
+
+
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fixtures/NoticeTokens.java]: There is an extra blank line here. Is there a reason for this?

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 5]
```diff
@@ -1540,9 +1542,14 @@ public void pageRequestNoticeIsSentWhenPolicyDefinesPageRequestNoticeConfigurati
...
     List<JsonObject> sentNotices = patronNoticesClient.getAll();
+
+    List<String> expectedContextPaths =
+      new ArrayList<>(NoticeTokens.EXPECTED_USER_TOKENS);
+    expectedContextPaths.addAll(NoticeTokens.EXPECTED_ITEM_TOKENS);
+    expectedContextPaths.addAll(NoticeTokens.EXPECTED_REQUEST_TOKENS);
     MatcherAssert.assertThat(sentNotices,
       hasItems(
-        equalsToEmailPatronNotice(requester.getId(), pageConfirmationTemplateId)));
+        equalsToEmailPatronNotice(requester.getId(), pageConfirmationTemplateId, expectedContextPaths)));
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/RequestsAPICreationTests.java]: This appears to check that the notice contains the expected contextual properties. Do any of the tests check that the values of those properties are what is expected?

File: src/test/java/api/support/fixtures/NoticeTokens.java (hunk 4 above)
Comment (dmytrotur) [reply src/test/java/api/support/fixtures/NoticeTokens.java]: No reason for this, removed

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 5 above)
Comment (dmytrotur) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: There are couple of tests where the due date property is checked. But most of the tests just check for not null.

Comment (marcjohnson-kint) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: How do we know that the notice is being sent with the correct context?

**Pull request merged**